### PR TITLE
daemon: Check that DaemonConfig is not changed after being published

### PR DESCRIPTION
**Pull request by jrajahalme**
Add sha256 sum to config and validate no change in immutable parts after DaemonConfig is resolved.

This has been carved out from #32981 to focus reviews.

### CODE REVIEW

File: pkg/option/config.go [hunk 1]
```diff
@@ -3941,16 +3952,18 @@ func (c *DaemonConfig) KubeProxyReplacementFullyEnabled() bool {
...
 // name 'daemon-config.json'. If this file already exists, it is renamed to
 // 'daemon-config-1.json', if 'daemon-config-1.json' also exists,
 // 'daemon-config-1.json' is renamed to 'daemon-config-2.json'
 func (c *DaemonConfig) StoreInFile(dir string) error {
-	backupFileNames := []string{
-		"agent-runtime-config.json",
-		"agent-runtime-config-1.json",
-		"agent-runtime-config-2.json",
-	}
+	c.shaSum = c.checksum()
```
Comment (asauber) [inline pkg/option/config.go:3966]: There is potential for a race here, the hash should be taken after the file is successfully written.

File: pkg/option/config.go [hunk 2]
```diff
@@ -3962,6 +3975,70 @@ func (c *DaemonConfig) StoreInFile(dir string) error {
...
+
+	var config DaemonConfig
+	err = json.Unmarshal(fileBytes, &config)
+
+	var diff string
+	if err != nil {
+		diff = fmt.Errorf("unmarshal failed %q: %w", string(fileBytes), err).Error()
+	} else {
+		// from https://github.com/google/go-cmp/issues/313#issuecomment-1315651560
+		opts := cmp.FilterPath(func(p cmp.Path) bool {
```
Comment (asauber) [inline pkg/option/config.go:4026]: Can we include a comment here that the purpose of this filter is to discover the exported field names during the comparison?

File: pkg/option/config.go [hunk 3]
```diff
@@ -3962,6 +3975,70 @@ func (c *DaemonConfig) StoreInFile(dir string) error {
...
+			if !ok {
+				return false
+			}
+			r, _ := utf8.DecodeRuneInString(sf.Name())
+			return !unicode.IsUpper(r)
+		}, cmp.Ignore())
+
+		diff = cmp.Diff(&config, c, opts,
+			cmpopts.IgnoreTypes(&IntOptions{}),
+			cmpopts.IgnoreTypes(&OptionLibrary{}))
```
Comment (asauber) [inline pkg/option/config.go:4037]: Why is it safe to ignore these fields?

**Review by asauber — CHANGES_REQUESTED** [review]

**Review by lmb — COMMENTED** [review]
Can you update the PR description to include what the problem you're solving is? Right now there is very little context.

**Comment by jrajahalme** [comment]
This got fixed and merged as part of #32981

**Comment by jrajahalme** [comment]
This got fixed and merged as part of https://github.com/cilium/cilium/pull/32981

**Pull request closed without merging**